docker/build_scripts/build.sh

```diff
@@ -65,23 +56,24 @@ yum -y update
 
 # EPEL support
 yum -y install wget curl
-# https://dl.fedoraproject.org/pub/epel/5/x86_64/epel-release-5-4.noarch.rpm
-cp $MY_DIR/epel-release-5-4.noarch.rpm .
-check_sha256sum epel-release-5-4.noarch.rpm $EPEL_RPM_HASH
+# https://dl.fedoraproject.org/pub/epel/6/x86_64/epel-release-6-8.noarch.rpm
+cp $MY_DIR/epel-release-6-8.noarch.rpm .
+check_sha256sum epel-release-6-8.noarch.rpm $EPEL_RPM_HASH
 
 # Dev toolset (for LLVM and other projects requiring C++11 support)
 curl -fsSLO http://people.centos.org/tru/devtools-2/devtools-2.repo
 check_sha256sum devtools-2.repo $DEVTOOLS_HASH
 mv devtools-2.repo /etc/yum.repos.d/devtools-2.repo
```

Comment:
Yes! `devtoolset-2` is why this has the same version of `gcc` as `manylinux1`, about which you were rightly suspicious. 

I think the change should be done in this PR, because upgrading to `devtoolset-7 will` [bump the version of `gcc` et. al. to 7.2.1](http://linux.web.cern.ch/linux/scientific6/docs/softwarecollections.shtml#devtoolset-7), but I've [advertised 4.8.2 in the PEP](https://github.com/markrwilliams/peps/blob/990c2636dcd70f53f795910ee2f7cbb7ca8d84d3/pep-9999.txt#L172-L173).  Installing `devtoolset-7` doesn't imply an upgrade to `glibc`, so I don't think it will result in incompatible executables, but I'll verify this with some additional tests.

Reply:
Unfortunately `devtoolset-7` hasn't been built for i686.  I can build it from scratch, much like the glibc patching, but I have concerns that it might inadvertently bump the libgcc version and thus require a change in the PEP.  I'm still investigating, though.

Reply:
Building `devtoolset-7` by hand seems like a lot of work. 32-bit linux users are *extremely* rare at this point – a quick check of bigquery says that in 2018, 0.7% of numpy's manylinux downloads were i686, and it's 0.2% for lxml, 0.8% for cryptography. (I'm spot-checking individual projects b/c I know they have both kinds of wheels, to avoid biases from projects that only have one or the other). If i686 has to use an earlier devtoolset it's probably fine, or if maintenance becomes too much of a hassle we could even drop support entirely.

Do you happen to know if any of the newer devtoolset releases target i686?

------

```
SELECT
  COUNT(*) AS downloads,
  REGEXP_EXTRACT(file.filename, r'.*-(manylinux.*)\.whl') AS manylinux_variant
FROM
  TABLE_DATE_RANGE([the-psf:pypi.downloads], TIMESTAMP("20180101"), CURRENT_TIMESTAMP())
WHERE
  file.project = 'cryptography'
GROUP BY
  manylinux_variant
ORDER BY
  downloads DESC
LIMIT
  1000
```

Reply:
Wheels are mostly indifferent to the C++ ABI, but what you say sounds plausible to me...

Reply:
To be clear, I'm not advocating for a change.  Just raising the question, because it is something we have needed to tackle at Anaconda.  We opted to change to the new ABI, because it is required for compiling C++14 and C++17 code.  Thus, we also have to ensure that users have a new enough libstdc++ available.  I'm not sure if the implication here is that devtoolset 7 can't compile such code, or if Red Hat has workarounds.

Reply:
@njsmith Your BigQuery query has convinced me that it's low risk to eliminate i686.  I'm going to ask for feedback on distutils-sig before removing it.

@msarahan Your understanding of how devtoolset 7 affects the C++ ABI matches mine (courtesy of @geofft).  Specifically, devtoolset-7 has the following linker script in place of `libstdc++`:

```
# cat /opt/rh/devtoolset-7/root/usr/lib/gcc/x86_64-redhat-linux/7/libstdc++.so
/* GNU ld script
   Use the shared library, but some functions are only in
   the static library, so try that secondarily.  */
OUTPUT_FORMAT(elf64-x86-64)
INPUT ( /usr/lib64/libstdc++.so.6 -lstdc++_nonshared )
```
In other words, the linker will indeed load all symbols it can from the system's `libstdc++` before statically linking in the remainder from its internal `libstdc++_nonshared` archive.

As a result upgrading to devtoolset-7 should keep the C++ ABI at the version described in [PEP 571](https://www.python.org/dev/peps/pep-0571/#the-manylinux2-policy).  This is also true of `libgcc` and its `GCC` versioned symbols.

Reply:
There has already been some discussion of 32-bit support [here](https://github.com/pypa/manylinux/issues/128).

Reply:
slightly off-topic, but i hope the devtoolset-7 repo signs packages using GPG? alternatively, we might be able to use their docker images

Reply:
@trishankatdatadog A good idea, but it looks like they only provide `devtoolset-7` images for CentOS 7 :(

Reply:
@markrwilliams looks like the SCL repo is GPG-signed, should be OK: https://copr.fedorainfracloud.org/coprs/rhscl/centos-release-scl/repo/epel-6/rhscl-centos-release-scl-epel-6.repo%20sudo%20yum%20install%20centos-release-scl

Reply:
@njsmith It looks like [it's not possible to build 32-bit artifacts on a 64-bit CentOS 6 image with devtoolset-7](https://github.com/pypa/manylinux/issues/128#issuecomment-365711771) either.  What if we ship devtoolset-7 in the 64-bit image and leave the i686 image with `devtoolset-2` from tru's repository?  The compiled output should be identical between the two devtoolsets, except that devtoolset-2 probably produces slower code.  That way we don't have to disadvantage projects like [lepton](https://pypi.org/project/lepton/) with relatively popular 32-bit `manylinux` wheels _or_ projects like `cryptography` and `numpy` whose `manylinux` downloads consist almost entirely of 64-bit wheels.

Reply:
It's true that `lepton` isn't popular, but they get value out of `manylinux` and I don't want to cast them adrift if possible.

It's also true that none of the devtoolsets are built for i686 for some reason.  I suspect that they're part of RHEL's paid offering.

I think it may be easy (in terms of build scripts, not computing resources) to build an `i686` version of devtoolset-7 with [mock](https://github.com/rpm-software-management/mock/wiki).

But the easiest thing I can think of is to leave the `i686` image with an older version of devtoolset and install devtoolset-7 on the `x86_64` image.  Most users would get a modern C++; people who don't need modern C++ but do want to build 32-bit artifacts -- in other words, the current users of `manylinux1`'s `i686` image -- could still do their thing; and anybody else could submit a PR :)

It's true that this would result in inconsistent _build_ environments, but I don't think it will result in meaningfully inconsistent _artifacts_ because of how the devtoolsets link stuff in.

Reply:
cryptography is down to 0.4% in the last 30 days for i686 wheels btw.